charon/src/ast/expressions_utils.rs

```diff
@@ -19,3 +20,93 @@ impl BorrowKind {
         }
     }
 }
+
+impl Place {
+    /// Compute the type of a place.
+    pub fn compute_projected_type(
+        &self,
+        type_decls: &Vector<TypeDeclId, TypeDecl>,
+        locals: &Vector<VarId, Var>,
+    ) -> Result<Ty, ()> {
+        // Lookup the local
+        let mut cur_ty = locals.get(self.var_id).ok_or(())?.ty.clone();
+
+        // Apply the projection
+        use ProjectionElem::*;
+        for pe in &self.projection {
+            cur_ty = match pe {
+                Deref => {
+                    use TyKind::*;
+                    match cur_ty.kind() {
+                        Ref(_, ty, _) | RawPtr(ty, _) => ty.clone(),
+                        Adt(..) | TypeVar(_) | Literal(_) | Never | TraitType(..) | DynTrait(_)
+                        | Arrow(..) => {
+                            // Unreachable
+                            return Err(());
+                        }
+                    }
+                }
+                Field(pkind, field_id) => {
+                    // Lookup the type decl
+                    use FieldProjKind::*;
+                    match pkind {
+                        Adt(type_decl_id, variant_id) => {
+                            let type_decl = type_decls.get(*type_decl_id).ok_or(())?;
```

Comment:
Hm, this case is annoying: the type decl may be missing because it was marked `--exclude`, which means this function can fail for reasons other than us messing up. It's fine for the current usage anyhow

Reply:
Let's just ignore that for now

Reply:
Yes, that's the reason why I put an `ok_or` rather than an `unwrap`